src/common/webrequest_curl.cpp

```diff
@@ -362,217 +510,1194 @@ void wxWebAuthChallengeCURL::SetCredentials(const wxWebCredentials& cred)
+                                const wxString& url,
+                                int id)
+{
+    // Allocate our handle on demand.
+    if ( !m_handle )
+    {
+        m_handle = curl_multi_init();
+        if ( !m_handle )
+        {
+            wxLogDebug("curl_multi_init() failed");
+            return wxWebRequestImplPtr();
+        }
+        else
+        {
+            curl_multi_setopt(m_handle, CURLMOPT_SOCKETDATA, this);
+            curl_multi_setopt(m_handle, CURLMOPT_SOCKETFUNCTION, SocketCallback);
+            curl_multi_setopt(m_handle, CURLMOPT_TIMERDATA, this);
+            curl_multi_setopt(m_handle, CURLMOPT_TIMERFUNCTION, TimerCallback);
+        }
+    }
+
+    return wxWebRequestImplPtr(new wxWebRequestCURL(session, *this, handler, url, id));
+}
+
+bool wxWebSessionCURL::StartRequest(wxWebRequestCURL & request)
+{
+    // Add request easy handle to multi handle
+    CURL* curl = request.GetHandle();
+    int code = curl_multi_add_handle(m_handle, curl);
+
+    if ( code == CURLM_OK )
+    {
+        request.SetState(wxWebRequest::State_Active);
+        m_activeTransfers[curl] = &request;
+
+        // Report a timeout to curl to initiate this transfer.
+        int runningHandles;
+        curl_multi_socket_action(m_handle, CURL_SOCKET_TIMEOUT, 0,
+                                 &runningHandles);
+
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+void wxWebSessionCURL::RequestHasTerminated(wxWebRequestCURL* request)
+{
+    CURL* curl = request->GetHandle();
+    TransferSet::iterator it = m_activeTransfers.find(curl);
+
+    if ( it != m_activeTransfers.end() )
+    {
+        // The transfer the CURL handle is performing is still in progress, but
+        // the web request object it belongs to is being deleted. Since the
+        // next step will call curl_easy_cleanup and any calls on the CURL
+        // handle after cleanup are illegal, remove it from the CURLM
+        // multihandle now.
+        curl_multi_remove_handle(m_handle, curl);
+        m_activeTransfers.erase(it);
+    }
+
+    curl_easy_cleanup(curl);
+}
+
+wxVersionInfo  wxWebSessionCURL::GetLibraryVersionInfo()
+{
+    const curl_version_info_data* vi = curl_version_info(CURLVERSION_NOW);
+    wxString desc = wxString::Format("libcurl/%s", vi->version);
+    if (vi->ssl_version[0])
+        desc += " " + wxString(vi->ssl_version);
+    return wxVersionInfo("libcurl",
+        vi->version_num >> 16 & 0xff,
+        vi->version_num >> 8 & 0xff,
+        vi->version_num & 0xff,
+        desc);
+}
+
+bool wxWebSessionCURL::CurlRuntimeAtLeastVersion(unsigned char major,
+                                                 unsigned char minor,
+                                                 unsigned char patch)
+{
+    unsigned int queryVersion = major;
+    queryVersion <<= 8;
+    queryVersion |= minor;
+    queryVersion <<= 8;
+    queryVersion |= patch;
+    return (ms_runtimeVersion >= queryVersion);
```

Comment:
It seems to be the same as
```suggestion
    return ms_runtimeVersion >= CURL_VERSION_BITS(major, minor, patch);
```

Reply:
I made this change.  The function originally took unsigned chars for which the shift would just zero out the chars.